src/ir/auth_logic/ast_construction_test.cc

```diff
@@ -0,0 +1,90 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//-------------------------------------------------------------------------------
+
+#include "src/ir/auth_logic/ast_construction.h"
+
+#include <iostream>
+#include <memory>
+#include <optional>
+#include <stdexcept>
+#include <string>
+#include <utility>
+#include <variant>
+#include <vector>
+
+#include "src/common/testing/gtest.h"
+#include "src/ir/auth_logic/ast.h"
+#include "src/ir/datalog/program.h"
+
+using testing::ElementsAre;
+using testing::Property;
+using testing::ResultOf;
+
+namespace raksha::ir::auth_logic {
+
+std::string GetQueryPrincipalName(const Query &query) {
+  return query.principal().name();
+}
+std::string GetQueryFactName(const Query &query) {
+  return std::get<datalog::Predicate>(
+             std::get<BaseFact>(query.fact().GetValue()).GetValue())
+      .name();
+}
+datalog::Predicate GetQueryFactArgs(const Query &query) {
+  return std::get<datalog::Predicate>(
+      std::get<BaseFact>(query.fact().GetValue()).GetValue());
+}
+
+struct QueryTestData {
+  std::string input_text;
+  std::string query_name;
+  std::string principal_name;
+  std::string fact_name;
+  datalog::Predicate predicate;
+};
+
+class AstConstructionTest : public testing::TestWithParam<QueryTestData> {};
+
+TEST_P(AstConstructionTest, SimpleTestWithQueryProgram) {
+  const auto &[input_text, query_name, principal_name, fact_name, predicate] =
+      GetParam();
+  Program prog_out = ParseProgram(input_text);
+  EXPECT_THAT(prog_out.queries(),
+              ElementsAre(Property(&Query::name, query_name)));
+  EXPECT_THAT(prog_out.queries(),
+              ElementsAre((ResultOf(GetQueryPrincipalName, principal_name))));
+  EXPECT_THAT(prog_out.queries(),
+              ElementsAre((ResultOf(GetQueryFactName, fact_name))));
+  EXPECT_THAT(prog_out.queries(),
+              ElementsAre((ResultOf(GetQueryFactArgs, predicate))));
+}
+
+INSTANTIATE_TEST_SUITE_P(
+    AstConstructionTest, AstConstructionTest,
+    testing::Values(
+        QueryTestData(
+            {R"(may_notifierA = query "UserA" says mayA("NotifierA"))",
+             "may_notifierA", R"("UserA")", "mayA",
+             datalog::Predicate("mayA", {R"("NotifierA")"},
+                                datalog::Sign::kPositive)}),
+        QueryTestData(
+            {R"(may_notifierB = query "UserB" says mayB("NotifierB"))",
+             "may_notifierB", R"("UserB")", "mayB",
+             datalog::Predicate("mayB", {R"("NotifierB")"},
+                                datalog::Sign::kPositive)})));
```

Comment:
I notice both of the tests test only the `kPositive` case. Should we also test the `kNegative` case?

Reply:
+1, if you are already handing `kNegative` in this PR.

Reply:
Modified test to include kNegative!